phoenix-core-client/src/main/java/org/apache/phoenix/filter/SkipScanFilter.java

```diff
@@ -576,13 +577,21 @@ private void setStartKey() {
 
     private int setStartKey(ImmutableBytesWritable ptr, int offset, int i, int nSlots, boolean atEndOfKey) {
         int length = ptr.getOffset() - offset;
-        startKey = copyKey(startKey, length + this.maxKeyLength, ptr.get(), offset, length);
+        startKey = copyKey(startKey, length + this.maxKeyLength + 1, ptr.get(), offset, length);
         startKeyLength = length;
         // Add separator byte if we're at end of the key, since trailing separator bytes are stripped
-        if (atEndOfKey && i > 0 && i-1 < nSlots) {
-            Field field = schema.getField(i-1);
+        if (atEndOfKey && i > 0 && i - 1 < nSlots) {
+            Field field = schema.getField(i - 1);
             if (!field.getDataType().isFixedWidth()) {
-                startKey[startKeyLength++] = SchemaUtil.getSeparatorByte(schema.rowKeyOrderOptimizable(), true, field);
+                if (field.getDataType() != PVarbinaryEncoded.INSTANCE) {
+                    startKey[startKeyLength++] =
+                        SchemaUtil.getSeparatorByte(schema.rowKeyOrderOptimizable(), true, field);
+                } else {
+                    byte[] sepBytes = SchemaUtil.getSeparatorBytesForVarBinaryEncoded(
+                        schema.rowKeyOrderOptimizable(), true, field.getSortOrder());
+                    startKey[startKeyLength++] = sepBytes[0];
+                    startKey[startKeyLength++] = sepBytes[1];
+                }
```

Comment:
Could we introduce an API which always returns a byte array and we simply iterate over the array. That should work regardless whether there is one separator byte or multiple and reduce the amount of code duplication.